elements/lisk-codec/src/json_wrapper.ts

```diff
@@ -0,0 +1,180 @@
+};
+
+const findObjectByPath = (
+	message: SchemaProps,
+	pathArr: string[],
+): SchemaProps | undefined => {
+	let result: SchemaProps = message;
+	for (let i = 0; i < pathArr.length; i += 1) {
+		if (!result.properties && !result.items) {
+			return undefined;
+		}
+		if (result.properties) {
+			result = result.properties[pathArr[i]];
+		} else if (result.items) {
+			const x = (result.items as SchemaProps).properties as SchemaPair;
+			result = x[pathArr[i]];
+		}
+	}
+	return result;
+};
+
+const isObject = (item: unknown): boolean =>
+	typeof item === 'object' &&
+	item !== null &&
+	!Array.isArray(item) &&
+	!Buffer.isBuffer(item);
+
+export const iterator = function iterator(
+	this: IteratableGenericObject,
+): {
+	next: () => {
+		done: boolean;
+		value: { value: IteratableGenericObject; key: string };
+	};
+} {
+	let index = 0;
+	const properties = Object.keys(this);
+	let Done = false;
+	return {
+		// eslint-disable-next-line @typescript-eslint/no-explicit-any
+		next: (): { done: boolean; value: { value: any; key: string } } => {
+			Done = index >= properties.length;
+			const obj = {
+				done: Done,
+				// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
+				value: { value: this[properties[index]], key: properties[index] },
+			};
+			index += 1;
+			return obj;
+		},
+	};
+};
+
+export const recursiveTypeCast = (
+	mode: 'toJSON' | 'fromJSON',
+	object: IteratableGenericObject,
+	schema: SchemaProps,
+	dataPath: string[],
+): void => {
+	for (const { key, value } of object) {
+		if (isObject(value)) {
+			dataPath.push(key);
+
+			(value as IteratableGenericObject)[Symbol.iterator] = iterator;
+			recursiveTypeCast(mode, value, schema, dataPath);
+			dataPath.pop();
+
+			delete (value as IteratableGenericObject)[
+				(Symbol.iterator as unknown) as string
+			];
+		} else if (Array.isArray(value)) {
+			dataPath.push(key);
+			const schemaProp = findObjectByPath(schema, dataPath);
+			if (schemaProp?.items?.type === 'object') {
+				for (let i = 0; i < value.length; i += 1) {
+					// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+					const arrayObject = value[i];
+
+					(arrayObject as IteratableGenericObject)[Symbol.iterator] = iterator;
+					recursiveTypeCast(mode, arrayObject, schema, dataPath);
+
+					delete (arrayObject as IteratableGenericObject)[Symbol.iterator];
+				}
+			} else {
+				for (let i = 0; i < value.length; i += 1) {
+					if (schemaProp === undefined || schemaProp.items === undefined) {
+						throw new Error(
+							`Invalid schema property found. Path: ${dataPath.join(',')}`,
+						);
+					}
+					if (mode === 'toJSON') {
+						// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
+						(object[key] as any)[i] = _liskMessageValueToJSONValue[
+							(schemaProp.items as SchemaScalarItem).dataType
+						](value[i]);
+					} else {
```

Comment:
This if/else inside loop and recursive calls does not look good and would be less performant. It would be nicer if we split the implementation into two functions.

Reply:
I get the point but if we split we end up with two functions that are exactly the same but in 1 line. Will check if possible to switch the map dynamically based on the parameter with little changes.

Reply:
But its inside loop and recursion. You yourself are emphasizing on the fact the even slightest performance really matters for `lisk-codec` then why not here.

Reply:
@nazarhussain I emphasise it for `encode()` and `decode()` this methods are extras.

Reply:
Changed to a map to make it constant time but still sorting out the type for the new map.

Reply:
Done. Resolving.